### PR TITLE
Changelog overlay

### LINKED ISSUE #2886: Changelog overlay
Now that the changelog page has been implemented on the website we can consider including it in the client.
twitterpost (taken from twitter)

- [ ] Build Header
header
If the json on the website is the one being returned it should look something like this:
headerjson
Important are:
  - display_version: obvious
  - users: this should probably be hidden when users == 0
  - is_featured: if the "panel" should be wide (?)
  - update_stream.display_name: obvious

  This object is inside an array which contains one for every build stream. The order of the array should probably match up with what's being shown.
~~**I'm unsure how the colour should be decided.**~~ Edit: Colours are going to be hardcoded for now until web returns them

- [ ] Chart
chart
~~Same problem as the header: How should the colours be decided?~~
I also don't know how this could be drawn, but it should probably be buffered.
- [ ] Changelog
changelog
The changelog json object is simlar to the one of the header but has an additional array containing the changes
changelogjson
  - type: the type of the change: fix (*), add (+) and probably misc (?)
  - category: categories are grouped
  - message_html: contains an additional message which also contain pictures, this somehow has to be converted to something usable in lazer

### PR DESCRIPTION
**Pull request by HoutarouOreki**
Mostly resolves #2886

Ready for reviews.

- [x] Header
- [x] Update Stream badges
- [x] ~Chart ([API works](https://streamable.com/1ger7))~
- [x] Changelog
- [x] Per-update stream changelog

[]()
- [x] ~Use no-break-space for `$" by {entry.GithubUser.DisplayName}"`~ not working
- [x] Match web sizing

ppy takeover
------------

- [x] Depends on https://github.com/ppy/osu-resources/pull/56.
- [x] Depends on #4830.
- [x] Depends on #4831.
- [x] Depends on #4832.
- [x] Depends on #4833.
- [x] Depends on #4834.
- [x] Depends on #4835.
- [x] Depends on #4837.
- [x] Depends on #4844.
- [x] Depends on ppy/osu-framework#2430.
- [x] Depends on ppy/osu-framework#2431.

---

### CODE REVIEW

File: osu.Game/Graphics/UserInterface/TooltipIconButton.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        public FontAwesome Icon
+        {
+            get { return icon.Icon; }
+            set { icon.Icon = value; }
+        }
+
+        public TooltipIconButton()
+        {
+            Children = new Drawable[]
+                {
```
Comment (AlFasGD) [inline osu.Game/Graphics/UserInterface/TooltipIconButton.cs]: Decrease indent

File: osu.Game/Overlays/Changelog/ChangelogStreams.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            BadgesContainer.OnUpdate = d =>
+            {
+                foreach (StreamBadge streamBadge in BadgesContainer.Children)
+                {
+                    streamBadge.OnActivation = () =>
+                    {
+                        SelectedRelease = streamBadge.ChangelogEntry;
+                        foreach (StreamBadge item in BadgesContainer.Children)
+                        {
+                            if (item.ChangelogEntry.Id != streamBadge.ChangelogEntry.Id) item.Deactivate();
```
Comment (AlFasGD) [inline osu.Game/Overlays/Changelog/ChangelogStreams.cs]: Too long statement; move to new line after the condition

File: osu.Game/Overlays/Changelog/ChangelogStreams.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            // how else can this be done?
+            BadgesContainer.OnUpdate = d =>
+            {
+                foreach (StreamBadge streamBadge in BadgesContainer.Children)
+                {
+                    streamBadge.OnActivation = () =>
+                    {
+                        SelectedRelease = streamBadge.ChangelogEntry;
+                        foreach (StreamBadge item in BadgesContainer.Children)
+                        {
```
Comment (AlFasGD) [inline osu.Game/Overlays/Changelog/ChangelogStreams.cs]: Brackets can be safely removed

File: osu.Game/Overlays/Changelog/Header/TextBadgePair.cs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        {
+            LineBadge.IsCollapsed = true;
+            Text.MoveToY(20, duration, easing)
+                .FadeOut(duration, easing);
+        }
+
+        public void ShowText(double duration = 0, string displayText = null, Easing easing = Easing.InOutCubic)
+        {
+            LineBadge.IsCollapsed = false;
+            if (!string.IsNullOrEmpty(displayText)) Text.Text = displayText;
```
Comment (AlFasGD) [inline osu.Game/Overlays/Changelog/Header/TextBadgePair.cs]: Use a new line after the condition

File: osu.Game/Online/API/Requests/Responses/APIChangelog.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        public long Users { get; set; }
+
+        [JsonProperty("is_featured")]
+        public bool IsFeatured { get; set; }
+
+        [JsonProperty("created_at")]
+        public DateTimeOffset CreatedAt { get; set; }
+
+        [JsonProperty("disqus_id")]
+        public string DisqusId { get; set; }
```
Comment (jorolf) [inline osu.Game/Online/API/Requests/Responses/APIChangelog.cs]: I doubt we'll ever support disqus inside the client

File: osu.Game/Overlays/Changelog/Header/TextBadgePairListing.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+        protected override bool OnHover(InputState state)
+        {
+            LineBadge.ResizeHeightTo(LineBadge.UncollapsedHeight, LineBadge.TransitionDuration, Easing.OutElastic);
+            return base.OnHover(state);
+        }
+
+        protected override void OnHoverLost(InputState state)
+        {
+            if (IsActivated == false) LineBadge.ResizeHeightTo(1, LineBadge.TransitionDuration, Easing.Out);
```
Comment (AlFasGD) [inline osu.Game/Overlays/Changelog/Header/TextBadgePairListing.cs]: Use `if (!IsActivated)` and new line after condition

File: osu.Game/Overlays/Changelog/ChangelogStreams.cs [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+        protected override bool OnHover(InputState state)
+        {
+            // is this nullreference-safe for badgesContainer?
+            foreach (StreamBadge streamBadge in BadgesContainer.Children)
+            {
+                if (SelectedRelease != null)
+                {
+                    if (SelectedRelease.UpdateStream.Id != streamBadge.ChangelogEntry.Id)
+                    {
```
Comment (AlFasGD) [inline osu.Game/Overlays/Changelog/ChangelogStreams.cs]: Brackets can be removed

File: osu.Game/Overlays/Changelog/ChangelogStreams.cs [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+                        streamBadge.Deactivate();
+                    }
+                }
+                else streamBadge.Deactivate();
```
Comment (AlFasGD) [inline osu.Game/Overlays/Changelog/ChangelogStreams.cs]: Move statement to new line

File: osu.Game/Overlays/Changelog/ChangelogStreams.cs [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+            }
+            return base.OnHover(state);
+        }
+
+        protected override void OnHoverLost(InputState state)
+        {
+            if (SelectedRelease == null)
+            {
+                foreach (StreamBadge streamBadge in BadgesContainer.Children) streamBadge.Activate(true);
```
Comment (AlFasGD) [inline osu.Game/Overlays/Changelog/ChangelogStreams.cs]: Use new line between loop and statement

File: osu.Game/Overlays/Changelog/ChangelogStreams.cs [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+                }
+                else streamBadge.Deactivate();
+            }
+            return base.OnHover(state);
+        }
+
+        protected override void OnHoverLost(InputState state)
+        {
+            if (SelectedRelease == null)
+            {
```
Comment (AlFasGD) [inline osu.Game/Overlays/Changelog/ChangelogStreams.cs]: Brackets can be safely removed

File: osu.Game/Overlays/Changelog/Header/TextBadgePair.cs [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+        {
+            IsActivated = true;
+            LineBadge.IsCollapsed = false;
+            Text.Font = "Exo2.0-Bold";
+            SampleActivate?.Play();
+        }
+
+        protected override bool OnHover(InputState state)
+        {
+            if (!IsActivated) sampleHover?.Play();
```
Comment (AlFasGD) [inline osu.Game/Overlays/Changelog/Header/TextBadgePair.cs]: Use new line after condition

File: osu.Game/Overlays/Changelog/Header/TextBadgePairRelease.cs [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+            Text.Alpha = 0;
+        }
+
+        public void SetText(string displayText) => Text.Text = displayText;
+
+        public void Activate(string displayText = null)
+        {
+            if (IsActivated)
+            {
+                if (displayText != Text.Text) ChangeText(transition_duration, displayText);
```
Comment (AlFasGD) [inline osu.Game/Overlays/Changelog/Header/TextBadgePairRelease.cs]: Use new line after condition

File: osu.Game/Overlays/Changelog/Header/TextBadgePairRelease.cs [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+            }
+            else
+            {
+                ShowText(transition_duration, displayText);
+                IsActivated = true;
```
Comment (AlFasGD) [inline osu.Game/Overlays/Changelog/Header/TextBadgePairRelease.cs]: The value will always end up being true, so setting it outside won't be a problem

File: osu.Game/Overlays/Changelog/StreamBadge.cs [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+                        new SpriteText
+                        {
+                            Text = ChangelogEntry.DisplayVersion,
+                            Font = @"Exo2.0-Light",
+                            TextSize = 21,
+                        },
+                        new SpriteText
+                        {
+                            Text = ChangelogEntry.Users > 0 ?
+                                string.Join(" ", ChangelogEntry.Users.ToString("N0"), "users online"):
```
Comment (AlFasGD) [inline osu.Game/Overlays/Changelog/StreamBadge.cs]: Prefer interpolated strings

File: osu.Game/Overlays/Changelog/StreamBadge.cs [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+                },
+            };
+        }
+
+        public void Activate(bool withoutHeaderUpdate = false)
+        {
+            isActivated = true;
+            this.FadeIn(transition_duration);
+            lineBadge.IsCollapsed = false;
+            if (!withoutHeaderUpdate) OnActivation?.Invoke();
```
Comment (AlFasGD) [inline osu.Game/Overlays/Changelog/StreamBadge.cs]: Use new line after condition

File: osu.Game/Overlays/ChangelogOverlay.cs [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+        {
+            switch (action)
+            {
+                case GlobalAction.Back:
+                    if (header.IsListingActivated()) State = Visibility.Hidden;
+
+                    // the problem here is that when hovering over the builds' container
+                    // and pressing back, they don't lower their opacity they're rehovered on
+                    else header.ActivateListing();
+                    return true;
```
Comment (AlFasGD) [inline osu.Game/Overlays/ChangelogOverlay.cs]: Can be removed

File: osu.Game/Overlays/ChangelogOverlay.cs [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+
+        // receive input outside our bounds so we can trigger a close event on ourselves.
+        public override bool ReceiveMouseInputAt(Vector2 screenSpacePos) => true;
+
+        public override bool OnPressed(GlobalAction action)
+        {
+            switch (action)
+            {
+                case GlobalAction.Back:
+                    if (header.IsListingActivated()) State = Visibility.Hidden;
```
Comment (AlFasGD) [inline osu.Game/Overlays/ChangelogOverlay.cs]: Use new line after condition

File: osu.Game/Overlays/ChangelogOverlay.cs [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+                content.Clear(); // this should probably happen with some transition
+                content.Add(new ChangelogContentGroup(Streams.SelectedRelease));
+            };
+            header.OnListingActivated += () =>
+            {
+                Streams.SelectedRelease = null;
+                content.Clear();
+                // should add listing to content here
+                if (!Streams.IsHovered)
+                    foreach (StreamBadge item in Streams.BadgesContainer.Children) item.Activate(true);
```
Comment (AlFasGD) [inline osu.Game/Overlays/ChangelogOverlay.cs]: Use new line between loop and statement

File: osu.Game/Overlays/ChangelogOverlay.cs [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+                else
+                    foreach (StreamBadge item in Streams.BadgesContainer.Children) item.Deactivate();
```
Comment (AlFasGD) [inline osu.Game/Overlays/ChangelogOverlay.cs]: Same as above

File: osu.Game/Overlays/ChangelogOverlay.cs [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+        }
+
+        public void FetchChangelog()
+        {
+            var req = new GetChangelogLatestBuildsRequest();
+            req.Success += res =>
+            {
+                Streams.BadgesContainer.Clear();
+                foreach (APIChangelog item in res)
+                {
```
Comment (AlFasGD) [inline osu.Game/Overlays/ChangelogOverlay.cs]: Brackets can be safely removed

**Review by Rekkonnect — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
Please learn to split out and name your commits better. They should describe what you are doing inside them.

File: osu.Game/Overlays/Changelog/StreamBadge.cs [hunk 21]
```diff
@@ -64,7 +64,7 @@ public StreamBadge(APIChangelog changelogEntry)
                         new SpriteText
                         {
                             Text = ChangelogEntry.Users > 0 ?
-                                string.Join(" ", ChangelogEntry.Users.ToString("N0"), "users online"):
+                                string.Format($"{ChangelogEntry.Users:N0} users online") :
```
Comment (AlFasGD) [inline osu.Game/Overlays/Changelog/StreamBadge.cs]: No need for `string.Format`, the interpolated string does the job by itself.

**Review by Rekkonnect — CHANGES_REQUESTED** [review]

File: osu.Game/Overlays/Changelog/StreamBadge.cs (hunk 21 above)
Comment (HoutarouOreki) [reply osu.Game/Overlays/Changelog/StreamBadge.cs]: Appveyor doesn't seem happy without it
image

Comment (HoutarouOreki) [reply osu.Game/Overlays/Changelog/StreamBadge.cs]: Should I ignore it?

Comment (AlFasGD) [reply osu.Game/Overlays/Changelog/StreamBadge.cs]: I have seen that error too, and I just specified the formatting culture, so using `.ToString("N0", CultureInfo.InvariantCulture)` is probably what it needs

Comment (HoutarouOreki) [reply osu.Game/Overlays/Changelog/StreamBadge.cs]: InvariantCulture breaks formatting for other languages. Right now I have that number nicely formatted in my language, using InvariantCulture sets everything to English. Are we doing this?

File: osu.Game/Graphics/UserInterface/LineBadge.cs [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+                    Width = height;
+                    Height = width;
+                }
+                else
+                    RelativeSizeAxes = value ? Axes.X : Axes.Y;
+                isHorizontal = value;
+            }
+        }
+
+        /// <summary>
```
Comment (peppy) [inline osu.Game/Graphics/UserInterface/LineBadge.cs]: xmldoc should be on class, not ctor

File: osu.Game/Online/API/Requests/Responses/APIChangelog.cs [hunk 23]
```diff
@@ -0,0 +1,120 @@
+﻿// Copyright (c) 2007-2018 ppy Pty Ltd <contact@ppy.sh>.
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+
+namespace osu.Game.Online.API.Requests.Responses
+{
+    public class APIChangelog
```
Comment (peppy) [inline osu.Game/Online/API/Requests/Responses/APIChangelog.cs]: shouldn't this be a `Build`?

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Graphics/UserInterface/LineBadge.cs (hunk 22 above)
Comment (peppy) [reply osu.Game/Graphics/UserInterface/LineBadge.cs]: (this applies to every case where you have descriptions of classes on the ctor)

**Comment by peppy** [comment]
Currently you have:
```
APIxxx <- Model
xxx <- Drawable
```

We generally use:
```
xxx <- Model
Drawablexxx <- Drawable
```

instead

**Comment by peppy** [comment]
Is this still going to get some attention?

**Comment by andy840119** [comment]
@Hanamuke Markdown has been merged into osu-framework
So if you have time you can continuous working on it 👍

**Comment by HoutarouOreki** [comment]
I tried a few days ago, but it seemed like Markdown Container doesn't
support Auto Size Axes, and that makes it kind of useless for this

**Comment by andy840119** [comment]
@HoutarouOreki is there any screenshot?
Or can you explain why it need screenshot?
Thanks : )

**Comment by peppy** [comment]
@HoutarouOreki I'm doing some work to bring this PR up-to-date and look at getting it merged. Just curious if you're still around, and also if you made further progress on the chart portion?

I'm going to tidy this up without the chart implementation for now, regarldess.

**Comment by HoutarouOreki** [comment]
Having final exams for a few more days. As for the chart, there was practically no further progress than what's shown in the video in the original post.

**Comment by peppy** [comment]
Thanks for the feedback. Could you also explain what the last checkbox (regex todo) refers to in the description?

**Comment by HoutarouOreki** [comment]
Ahh, that was supposed to help replace `TextSize` with the value from a comment (`// Web: xx,`) in bulk, by plugging it into Find&Replace in Visual Studio, but it doesn't work anymore since the way to change the font size has changed.

I was using different font sizes than these from the web to get a more 1:1 look on my monitor. For example, lazer's 28px font size is/was equivalent to 24px web's font size for me. Kept original sizes in comments to be able to later replace them without checking them through a browser.

gif

**Comment by peppy** [comment]
i see, thanks!

**Comment by peppy** [comment]
As a heads-up, the original code was such a mess I have considered multiple times starting from scratch. For now I will continue to work with this and refactor it, but there's still a long way to go.

I hope you can learn by watching your code be almost 100% rewritten from its original state.

Just a few things done wrong:
- Too many unnecessary bool flags and states
- Not using object oriented styles (every class has a `ShowListing` and `ShowBuild` instead of having two derived versions)
- Not using existing breadscrumb control (see `ScreenBreadcrumbControl`)
- No tests for smaller individual pieces (like a single changelog display)
- Not setup for dynamic compilation – I don't want to know how many times you rebuilt and restarted the app while working on this. Please see https://github.com/ppy/osu-framework/wiki/Development-and-Testing
- Rather than custom logic to show/hide listing/build and save scroll positions, it should probably be using a screen stack.
- Events and EventHandlers all over the place where they are not necessary.

**Comment by HoutarouOreki** [comment]
Most of these were apparent to me at the point of considering MarkdownContainer, though I didn't have an exact precise idea of how they should be done correctly for most of them. Didn't want to bother myself too much with it because of my lessened amount of free time, considering that there was *so* much wrong. Your rewrite will definitely give some answers. Appreciated. Now that I think about it I should've closed the pr or edited the op to let others know I'm not gonna work on it for some time.

File: osu.Game/Overlays/Changelog/ChangelogBuild.cs [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+
+            RelativeSizeAxes = Axes.X;
+            AutoSizeAxes = Axes.Y;
+            Direction = FillDirection.Vertical;
+            Padding = new MarginPadding { Horizontal = 70 };
+
+            Children = new Drawable[]
+            {
+                CreateHeader(),
+                ChangelogEntries = new FillFlowContainer
```
Comment (UselessToucan) [inline osu.Game/Overlays/Changelog/ChangelogBuild.cs]: I suggest to create a private method called `CreateList` or similar.

Comment (peppy) [reply osu.Game/Overlays/Changelog/ChangelogBuild.cs]: why?

Comment (peppy) [reply osu.Game/Overlays/Changelog/ChangelogBuild.cs]: you mean to move the parsing logic i haven't yet refactored? this PR is not ready for merge just yet.

**Comment by peppy** [comment]
@HoutarouOreki This is all done now. working your code into a good state probably took 2-3x what it would have been writing from scratch, but i hope looking through my commits (and the approach to how i went about fixing pieces) is helpful to you.

Make sure to check our all the pull requests and fixes which spawned from this implementation (added to the original description of this PR).

**Comment by smoogipoo** [comment]
@peppy do a visual pass on it once more and check my recent commits.

**Review by smoogipoo — APPROVED** [review]

**Review by peppy — APPROVED** [review]

**Pull request merged**